data_model/derive/src/event_set.rs

```diff
@@ -0,0 +1,395 @@
+                let flag_name = flag_ident.to_string();
+                quote! {
+                    #flag_name
+                }
+            })
+            .collect::<Vec<_>>();
+        // names of all the flags in `quotes` to use in error message
+        let flag_names_str = variants
+            .iter()
+            .map(|EventSetVariant { flag_ident, .. }| format!("`{flag_ident}`"))
+            .collect::<Vec<_>>()
+            .join(", ");
+        // patterns for matching events in the `matches` method
+        let event_patterns = variants.iter().map(
+            |EventSetVariant {
+                 event_ident,
+                 fields_style,
+                 ..
+             }| {
+                quote! {
+                    #event_enum_ident::#event_ident #fields_style
+                }
+            },
+        );
+
+        let doc = format!(" An event set for [`{event_enum_ident}`]s\n\nEvent sets of the same type can be combined with a custom `|` operator");
+
+        tokens.extend(quote! {
+            #[derive(
+                Copy,
+                Clone,
+                PartialEq,
+                Eq,
+                PartialOrd,
+                Ord,
+                Hash,
+                // this introduces tight coupling with these crates
+                // but it's the easiest way to make sure those traits are implemented
+                parity_scale_codec::Decode,
+                parity_scale_codec::Encode,
+                // TODO: we probably want to represent the bit values for each variant in the schema
+                iroha_schema::IntoSchema,
+            )]
+            #[repr(transparent)]
+            #[doc = #doc]
+            #vis struct #set_ident(u32);
+
+            // we want to imitate an enum here, so not using the SCREAMING_SNAKE_CASE here
+            #[allow(non_upper_case_globals)]
+            impl #set_ident {
+                #( #flag_defs )*
+            }
+
+            impl #set_ident {
+                /// Creates an empty event set
+                pub const fn empty() -> Self {
+                    Self(0)
+                }
+
+                /// Creates an event set containing all events
+                pub const fn all() -> Self {
+                    Self(
+                        #(
+                            #flag_idents.0
+                        )|*
+                    )
+                }
+
+                /// Combines two event sets by computing a union of two sets
+                ///
+                /// A const method version of the `|` operator
+                pub const fn or(self, other: Self) -> Self {
+                    Self(self.0 | other.0)
+                }
+
+                /// Checks whether an event set is a superset of another event set
+                ///
+                /// That is, whether `self` will match all events that `other` contains
+                pub const fn contains(&self, other: Self) -> bool {
```

Comment:
This might be useful for event filter introspection (and was the main reason I even included as a public API). If we don't want to serve the filter introspection use-case I guess this could just be made private :shrug:

Reply:
I prefer the keep-it-private until required otherwise mentality

Reply:
Ok, I'll hide it for now